### PR TITLE
[RFR] Ob add questions

### PR DESCRIPTION
**Pull request by SpencerCDixon**
Created and passed two tests (edit question and edit answer)

### CODE REVIEW

**Comment by SpencerCDixon** [comment]
This is actually Omid's pull request not mine. Dunno how we should handle this...

File: app/controllers/answers_controller.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+class AnswersController < ApplicationController
+  def create
```
Comment (dpickett) [inline app/controllers/answers_controller.rb]: I must be signed in to answer, right? We should have a `before_action` that ensures that.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -2,6 +2,7 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :null_session
+  add_flash_types :info
```
Comment (dpickett) [inline app/controllers/application_controller.rb]: what is the benefit of having an additional flash type in addition to `:notice` and `:alert`?

File: app/controllers/questions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
+class QuestionsController < ApplicationController
+  def index
+    @questions = Question.all
```
Comment (dpickett) [inline app/controllers/questions_controller.rb]: this data should be constrained or paginated

File: app/controllers/questions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  private
+
+  def create_params
+    params.require(:question).permit(:title, :body)
+  end
+
+  def update_params
+    params.require(:question).permit(:accepted_answer_id, :title, :body)
```
Comment (dpickett) [inline app/controllers/questions_controller.rb]: assigning the `accepted_answer_id` is a function or responsibility of another resource. Manage this through a `accepted_answers_controller` and a post or put to a url like `/questions/<id>/accepted_answer`

File: app/models/question.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+class Question < ActiveRecord::Base
+  belongs_to :user
+  belongs_to :accepted_answer, class_name: "Answer"
```
Comment (dpickett) [inline app/models/question.rb]: hard return on `class_name` argument

File: app/models/question.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  has_many :answers, dependent: :destroy
```
Comment (dpickett) [inline app/models/question.rb]: hard return on `dependent` argument

File: app/models/question.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+  validates :title, presence: true, length: { in: 10..200 }
+  validates :body, presence: true, length: { in: 15..10000 }
+  validates :user, presence: true
+
+  validate :accepted_answer_belongs_to_question
+
+  def accepted_answer_belongs_to_question
```
Comment (dpickett) [inline app/models/question.rb]: good! This should be a protected method as it should not be invoked externally

File: app/models/question.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    if accepted_answer && !answers.include?(accepted_answer)
+      errors.add(:accepted_answer_id, "must belong to this question")
+    end
+  end
+
+  def sorted_answers
+    answers.sort_by { |answer| answer.accepted? ? 0 : 1 }
```
Comment (dpickett) [inline app/models/question.rb]: utilize a scope leverage the power of SQL here

```ruby
question.answers.accepted_first
```

File: app/models/user.rb [hunk 9]
```diff
@@ -5,6 +5,8 @@ class User < ActiveRecord::Base
   has_many :teams, through: :team_memberships
   has_many :assignments, through: :teams
   has_many :announcements, through: :teams
+  has_many :questions
+  has_many :answers, through: :questions
```
Comment (dpickett) [inline app/models/user.rb]: hard return on the `through` argument

File: db/migrate/20141217191055_add_accepted_answer_id_to_questions.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class AddAcceptedAnswerIdToQuestions < ActiveRecord::Migration
+  def change
+    add_column :questions, :accepted_answer_id, :integer
```
Comment (dpickett) [inline db/migrate/20141217191055_add_accepted_answer_id_to_questions.rb]: we should index this if we'll be sorting or querying based on it.

File: spec/features/questions_spec.rb [hunk 11]
```diff
@@ -0,0 +1,177 @@
+require "rails_helper"
+
+feature "questions" do
+  scenario "view the newest questions" do
+    questions = FactoryGirl.create_list(:question, 3)
```
Comment (dpickett) [inline spec/features/questions_spec.rb]: is there tremendous benefit to creating more than 1 question for the purpose of this test? Each write to the database incurs a performance cost

File: spec/features/questions_spec.rb [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+    expect(page).to have_content("Newest Questions")
+
+    questions.each do |question|
+      expect(page).to have_link(question.title, href: question_path(question))
+    end
+  end
+
+  scenario "view unanswered questions"
+
+  scenario "view a question with no answers" do
```
Comment (dpickett) [inline spec/features/questions_spec.rb]: these scenarios likely belong in other files. viewing a list of questions and viewing an individual question feel like separate features to me

File: spec/features/questions_spec.rb [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+      click_button "Submit Answer"
+
+      expect(page).to have_content("Answer saved.")
+      expect(page).to have_content("You need to reticulate the splines.")
+      expect(page).to have_content("#{user.username} answered")
+
+      expect(Answer.count).to eq(1)
+    end
+
+    scenario "display error when submitting a blank answer" do
```
Comment (dpickett) [inline spec/features/questions_spec.rb]: answering a question belongs in a separate spec file

File: spec/features/questions_spec.rb [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+      question = FactoryGirl.create(:question)
+
+      visit question_path(question)
+      click_button "Submit Answer"
+
+      expect(page).to have_content("Failed to save answer.")
+      expect(Answer.count).to eq(0)
+    end
+
+    scenario "accept an answer" do
```
Comment (dpickett) [inline spec/features/questions_spec.rb]: accepting an answer belongs in a separate spec file

File: spec/features/questions_spec.rb [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+      click_button "Accept Answer"
+
+      expect(page).to have_content("Your question has been updated.")
+      expect(page).to have_content("accepted answer")
+
+      question.reload
+      expect(question.accepted_answer).to eq(answer)
+    end
+
+    scenario "edit question" do
```
Comment (dpickett) [inline spec/features/questions_spec.rb]: editing a question belongs in a separate spec file

File: spec/features/questions_spec.rb [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+    end
+
+    scenario "delete question"
+    scenario "delete answer"
+    scenario "un-accept an answer"
+    scenario "change accepted answer"
+    scenario "comment on a question"
+    scenario "comment on an answer"
+
+    scenario "only original asker can accept answer" do
```
Comment (dpickett) [inline spec/features/questions_spec.rb]: this belongs contextually with accepting an answer.

**Comment by dpickett** [comment]
a good first pass

**Comment by atsheehan** [comment]
Merge into the `add-questions` at this point, not master.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (SpencerCDixon) [inline app/controllers/application_controller.rb]: It allows us to have helper methods in the redirects.  Usually I assign different styles to the type of flash and find it useful.  If you don't want it we can delete it though.

**Pull request closed without merging**